docs(client): launch post (#86)
kadena-community/kadena.js: packages/libs/bootstrap-lib/docs/launch-post.md

Title: Release @kadena/client

We've created a library that allows Javascript/Typescript users to easily
interact with the Kadena Blockchain. We've explicitly kept out creating
contracts as it's way more complicated to transpile Javascript to Pact, as

> contracts need to be Pact lang in the end anyway.

Reviewer: ```suggestion
contracts need to be written in Pact anyway.
```
?

> contracts need to be Pact lang in the end anyway.

Author: We're not going to build stuff to make smart contracts from javascript. At least not in the forseeable future

> contracts need to be Pact lang in the end anyway.

Reviewer 2: > Note: This release does not cover smart contract creation yet, but that is on our roadmap.
I rather not give commitments about items that may change in the future

  - [Load contracts from the blockchain](#load-contracts-from-the-blockchain)
  - [Manually singing the transaction](#manually-singing-the-transaction)
  - [Automated sign request to **Chainweaver desktop**](#automated-sign-request-to-chainweaver-desktop)
- [Building a simple transaction using a template](#building-a-simple-transaction-using-a-template)
  - [Load the contract repository](#load-the-contract-repository)
  - [Generate code from templates](#generate-code-from-templates)
  - [A function is generated from a template](#a-function-is-generated-from-a-template)
  - [Using the generated code](#using-the-generated-code)
- [Further development](#further-development)
- [Contact the team](#contact-the-team)

# Prerequisites

We're not going over installation of nodejs and package management, this
information is widely available on the internet. We will provide the bare
minimum of the things you need to do in order to start using @kadena/client:

- install nodejs version 14.x or 16.x
- create a directory, bootstrap a `package.json` by running `npm init` or
  `npm init -y` to use defaults
- install the client `npm install @kadena/client`

> - install the commandline tool `npm install @kadena/pactjs-cli`

Reviewer 2: should we add `--save-dev`

Author: That's by default

Reviewer 3: Good call actually, the pactjs cli could be incorporated in a prod workflow somehow and break if we suggest save-dev

Doesn't save much of anything imho anyway

# Bootstrapping @kadena/client

We wanted @kadena/client to be self sufficient even when you're not using
default contracts like `coin` or `marmalade`. That's why you have to generate
the interfaces that are used by `@kadena/client` from information on the

> blockchain or from your own smart contracts.

Reviewer 3: If I understand correctly, this is necessary for coin and marmelade as well (if not, it should be stated explicitly - "Skip this if you aren't using non-standard smart contracts)

Should we include those commonly used contracts' interfaces and include them somehow? It would remove some friction - having a simple hello-world or pay-txn up and running with the least steps possible would be better

Counterpoint: "What about when coin/marmelade is updated?"

-    Updates are always backwards compatible with respect to the interface IIRC, so existing users would not see a regression
-    We would update the upstream repo and new users could get the full functionality
-    We should include some script to refresh those definitions from chain

Reviewer 3: > I'm thinking about leaving out marmalade alltogether as it's still in development and doesn't really add value mentioning it and only adds confusion

+1

> I want to have a kadena-templates repo that contains basic stuff for doing transactions with coin contract. Similar to the one from kda-tool.

Great!

Options:  
 -m, --module <module>  
 The module you want to retrieve (e.g. "coin")

-o, --out <file> File to write the contract

-n, --network <network> Network to retrieve from (default "mainnet") (default:
"mainnet")

-c, --chain <number> Chain to retrieve from (default 1) (default: 1)

-h, --help display help for command

````

</details>

## Generate interfaces

Using the contract we'll now generate all the functions (`defun`s) with their
(typed) arguments and the capabilities (`defcap`s).

```bash
pactjs contract-generate --file "./contracts/coin.module.pact"
````

The log shows what has happened. Inside the `node_modules` directory, a new
package has been created: `@kadena/generated`. This package is referenced by
`@kadena/client` to give you type information.

# Building a simple transaction from the contract

Now that everything is bootstrapped, we can start building transactions.

Create a new file and name it `transfer.ts` (or `.js`).

```ts
import { Pactjs } from '@kadena/client';

// store the builder in a variable
const transactionBuilder =
  // basic call results in Pact code `(coin.transfer "k:your-pubkey" "k:receiver-pubkey" 231.0)`
  Pact.modules.coin

>     .transfer('k:your-pubkey', 'k:receiver-pubkey', 231)

Reviewer 2: Should we add PactNumber here already?

    // add necessary coin.GAS capability (this defines who pays the gas)
    .addCap('coin.GAS', 'your-pubkey')
    // add necessary coin.TRANSFER capability
    .addCap(
      'coin.TRANSFER',
      'your-pubkey',
      'k:your-pubkey',
      'k:receiver-pubkey',
      231,
    )
    // the minimum you NEED to add is the sender of this transaction
    .setMeta({
      sender: 'your-pubkey',
    });
```

> Take note of the following:

Reviewer 2: should we use the note option here?
> **Note**
> text

> ```
> 
> Take note of the following:
> 
> - namespaced arguments (`k:`, `w:` etc) are account-names, where non-namespaced
>   arguments are public-keys
> - the contract doesn't specify whether you need to pass an **account-name** or
>   **public-key**. This is knowledge that can be obtained by inspecting the
>   contract downloaded earlier or consulting the documentation for the contract.
> - `addCap` function accepts a `capability` and a `public-key` of the signer of
>   the capability. The other arguments are defined by the contract. `coin.GAS`
>   doesn't have any arguments, `coin.TRANSFER` does.
> - `setMeta`s object has a `sender` property, which is a `public-key`.

Reviewer 2: ```suggestion
```

> **Note** 
> Take note of the following:
> - namespaced arguments (`k:`, `w:` etc) are account-names, where non-namespaced
  arguments are public-keys
> - the contract doesn't specify whether you need to pass an **account-name** or
  **public-key**. This is knowledge that can be obtained by inspecting the
  contract downloaded earlier or consulting the documentation for the contract.
> - `addCap` function accepts a `capability` and a `public-key` of the signer of
  the capability. The other arguments are defined by the contract. `coin.GAS`
  doesn't have any arguments, `coin.TRANSFER` does.
> - `setMeta`s object has a `sender` property, which is a `public-key`.
```

> ```
> 
> Take note of the following:
> 
> - namespaced arguments (`k:`, `w:` etc) are account-names, where non-namespaced
>   arguments are public-keys
> - the contract doesn't specify whether you need to pass an **account-name** or
>   **public-key**. This is knowledge that can be obtained by inspecting the
>   contract downloaded earlier or consulting the documentation for the contract.
> - `addCap` function accepts a `capability` and a `public-key` of the signer of
>   the capability. The other arguments are defined by the contract. `coin.GAS`
>   doesn't have any arguments, `coin.TRANSFER` does.
> - `setMeta`s object has a `sender` property, which is a `public-key`.

Reviewer 2: ```suggestion
```
> **Note** 
> Take note of the following:
> - namespaced arguments (`k:`, `w:` etc) are account-names, where non-namespaced
  arguments are public-keys
> - the contract doesn't specify whether you need to pass an **account-name** or
  **public-key**. This is knowledge that can be obtained by inspecting the
  contract downloaded earlier or consulting the documentation for the contract.
> - `addCap` function accepts a `capability` and a `public-key` of the signer of
  the capability. The other arguments are defined by the contract. `coin.GAS`
  doesn't have any arguments, `coin.TRANSFER` does.
> - `setMeta`s object has a `sender` property, which is a `public-key`.
```

## Manually singing the transaction

To sign the transaction, you can use the builder to output something that can be
pasted into the `SigData` of Chainweaver.

```ts
// createTransaction() will calculate hashes and finalizes the unsigned transaction
const unsignedTransaction = transactionBuilder.createTransaction();

console.log(JSON.stringify(unsignedTransaction));
```

## Automated sign request to **Chainweaver desktop**

> This is not very user friendly. Using the `transaction` we can send a

Reviewer: I would move this sentence to the end of the previous paragraph, so it's clear that manually signing isn't very user friendly (but there is an alternative below)

> This is not very user friendly. Using the `transaction` we can send a

Reviewer 2: Remove the part "This is not very user friendly".
We know it is and it is something we are going to fix indeed as stated below

sign-request to Chainweaver. **(NB: this can only with the desktop version, not
the web-version, as it's
[exposing port 9467](https://kadena-io.github.io/signing-api/)**

> > In the future we will provide an interface with WalletConnect. This is not yet

Reviewer 2: add note here as well

> > In the future we will provide an interface with WalletConnect. This is not yet
> > finalized. Once it is, we'll update the `@kadena/client` accordingly

Reviewer 2: ```suggestion
> **Note**
> In the future, we will provide an interface with WalletConnect. This is not yet
> finalized. Once it is, we'll update the `@kadena/client` accordingly
```

```ts
import {
  signAndSubmitWithChainweaver,
  signWithChainweaver,
} from '@kadena/client';

// pass the transactionBuilder object, as metadata can still be changed
const submitResult = signAndSubmitWithChainweaver(transactionBuilder);

// use the finalized transaction, and sign it with Chainweaver
cont signedTransaction = signWithChainweaver(unsignedTransaction)
  .then(console.log)
  .catch(console.error);

> ```

Reviewer 2: yesss

Take note of the following:

- `createTransaction()` will **finalize** the transaction. The hash will be
  calculated. Things like `sender`, `gasPrice` or `gasLimit` **CANNOT be
  changed** anymore.
- `signWithChainweaver` needs the finalized transaction
- `signAndSubmitWithChainweaver` needs the "open" transaction, as it needs to
  calculate the hash for the metadata (sender, gas-parameters)

# Building a simple transaction using a template

To provide contract-developers a way to communicate how their contracts should

> be used, we added a way to get intelliSense for templates. Contract-developers

Reviewer: ```suggestion
be used, we added a way to get autocompletion for templates. Contract-developers
```

- `-f, --file`
  - selecting a file as input will create ONLY code for that file
  - selecting a directory as input will create code for ALL the templates in the
    directory
- `-o, --out`
  - when the output is a file, the code for the templates will end up in that
    file
  - when the output is a directory, an `index.ts` will be created in that
    directory, containing the code for the templates

## A function is generated from a template

Each file in the repository is converted to a function that can be called. The
function has one argument; an object that contains named key-value pairs for
each variable in the template.

For example, a bogus template that looks like this

```txt
# ./hello.txt
This is a Hello, {{name}}!
```

Will have it's function call:

```ts
import myTemplates from './myTemplates';

myTemplates.hello({ name: 'alber70g' });
```

Of course this isn't a valid template to be used as a transaction, so this won't
work. But this outlines the general idea of how templates are used.

## Using the generated code

Let's say we're using this template. Templates **aren't** valid `yaml`. They are
however checked to be valid transactions when used as templates.

> ```

Reviewer 2: add ```yaml

> ```

Reviewer 3: ```suggestion
```yaml
```

> ```
> code: |-
>   (coin.transfer "{{fromAcct}}" "{{toAcct}}" {{amount}})
> data:
> publicMeta:
>   chainId: "{{chain}}"
>   sender: {{fromAcct}}
>   gasLimit: 2500
>   gasPrice: 1.0e-8
>   ttl: 600
> networkId: {{network}}
> signers:
>   - pubKey: {{fromKey}}
>     caps:
>       - name: "coin.TRANSFER"
>         args: [{{fromAcct}}, {{toAcct}}, {{amount}}]
>       - name: "coin.GAS"
>         args: []
> type: exec
> ```

Reviewer 2: ```suggestion
```yaml
code: |-
  (coin.transfer "{{fromAcct}}" "{{toAcct}}" {{amount}})
data:
publicMeta:
  chainId: "{{chain}}"
  sender: {{fromAcct}}
  gasLimit: 2500
  gasPrice: 1.0e-8
  ttl: 600
networkId: {{network}}
signers:
  - pubKey: {{fromKey}}
    caps:
      - name: "coin.TRANSFER"
        args: [{{fromAcct}}, {{toAcct}}, {{amount}}]
      - name: "coin.GAS"
        args: []
type: exec
```
```
